test/functional/wallet_upgradewallet.py

```diff
@@ -0,0 +1,155 @@
+
+contrib/devtools/previous_release.sh -b v0.15.2 v0.16.3
+"""
+
+import os
+import shutil
+
+from test_framework.test_framework import BitcoinTestFramework, SkipTest
+from test_framework.util import (
+    adjust_bitcoin_conf_for_pre_17,
+    assert_equal,
+    assert_greater_than,
+    assert_is_hex_string
+)
+
+class UpgradeWalletTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 3
+        self.extra_args = [
+            ["-addresstype=bech32"], # current wallet version
+            ["-usehd=1"],            # v0.16.3 wallet
+            ["-usehd=0"]             # v0.15.2 wallet
+        ]
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def setup_network(self):
+        self.setup_nodes()
+
+    def setup_nodes(self):
+        if os.getenv("TEST_PREVIOUS_RELEASES") == "false":
+            raise SkipTest("upgradewallet RPC tests")
+
+        releases_path = os.getenv("PREVIOUS_RELEASES_DIR") or os.getcwd() + "/releases"
+        if not os.path.isdir(releases_path):
+            if os.getenv("TEST_PREVIOUS_RELEASES") == "true":
+                raise AssertionError("TEST_PREVIOUS_RELEASES=1 but releases missing: " + releases_path)
+            raise SkipTest("This test requires binaries for previous releases")
+
+        self.add_nodes(self.num_nodes, extra_args=self.extra_args, versions=[
+            None,
+            160300,
+            150200
+        ], binary=[
+            self.options.bitcoind,
+            releases_path + "/v0.16.3/bin/bitcoind",
+            releases_path + "/v0.15.2/bin/bitcoind",
+        ], binary_cli=[
+            self.options.bitcoincli,
+            releases_path + "/v0.16.3/bin/bitcoin-cli",
+            releases_path + "/v0.15.2/bin/bitcoin-cli",
+        ])
+        # adapt bitcoin.conf, because older bitcoind's don't recognize config sections
+        adjust_bitcoin_conf_for_pre_17(self.nodes[1].bitcoinconf)
+        adjust_bitcoin_conf_for_pre_17(self.nodes[2].bitcoinconf)
+        self.start_nodes()
+
+    def dumb_sync_blocks(self):
+        """
+        Little helper to sync older wallets. 
+        Notice that v0.15.2's regtest is hardforked, so there is 
+        actually no sync for it. v0.15.2 is only being used
+        to test version upgrade and master hash key presence.
+        v0.16.3 is being used to test version upgrade and balances.
+        Further info: https://github.com/bitcoin/bitcoin/pull/18774#discussion_r416967844
+        """
+        node_from = self.nodes[0]
+        to_height = node_from.getblockheader(node_from.getbestblockhash())['height']
+        for n in range(1, len(self.nodes)):
+            height = self.nodes[n].getblockheader(self.nodes[n].getbestblockhash())['height']
+            for i in range(height, to_height+1):
+                b = node_from.getblock(blockhash=node_from.getblockhash(i), verbose=0)
+                self.nodes[n].submitblock(b)
+
```

Comment:
```suggestion
            assert_equal(self.nodes[n].getblockcount(), to_height)
```

Reply:
Thanks. Btw., have tried to add you as Co-Author, but couldn't find your email or "no-reply" email.